Avoid calling updateCompositingLayers() more than once before each render.

mattwoodrow opened the pull request:
#### b6b74556b8e6ecc165a8e1edb5fc73cc8e5853c2
<pre>
Avoid calling updateCompositingLayers() more than once before each render.
<a href="https://bugs.webkit.org/show_bug.cgi?id=84393">https://bugs.webkit.org/show_bug.cgi?id=84393</a>
&lt;<a href="https://rdar.apple.com/11359525">rdar://11359525</a>&gt;

Reviewed by Simon Fraser.

Calling updateCompositingLayers after each layout causes unnecessary work, since
the layers might change again before rendering.

This change defers this work until Page::updateRendering if possible, and improves
performance on pages that force layout flushes outside of rendering.

* LayoutTests/compositing/overflow/scrolling-without-painting.html:
* LayoutTests/compositing/overflow/updating-scrolling-content.html:
* LayoutTests/compositing/scrolling/touch-scroll-to-clip.html:
* LayoutTests/compositing/updates/no-style-change-updates.html:
* LayoutTests/editing/editable-region/iframe-expected.txt:
* LayoutTests/fast/css/transform-function-perspective-crash.html:
* LayoutTests/fast/scrolling/iframe-scrollable-after-back.html:
* LayoutTests/fast/scrolling/mac/scrollbars/overflow-overlay-scrollbar-hit-test.html:
* LayoutTests/fast/scrolling/mac/scrollbars/overflow-overlay-scrollbar-hovered.html:
* LayoutTests/inspector/layers/layerTreeDidChange.html:
* LayoutTests/platform/gtk/compositing/scrolling/touch-scroll-to-clip-expected.txt: Copied from LayoutTests/platform/ios/compositing/overflow/updating-scrolling-content-expected.txt.
* LayoutTests/platform/ios/compositing/overflow/updating-scrolling-content-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/fast/text/mark-matches-overflow-clip-expected.txt:
* LayoutTests/scrollingcoordinator/scrolling-tree/clipped-layer-in-overflow-nested.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/clipped-layer-in-overflow.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-inside-relative-inside-stacking-overflow-inside-transformed.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-inside-stacking-overflow-inside-transformed.html:
* LayoutTests/tiled-drawing/scrolling/clamp-out-of-bounds-scrolls.html:
* LayoutTests/tiled-drawing/scrolling/scrolling-tree-after-scroll.html:
* LayoutTests/tiled-drawing/scrolling/slow-scrolling.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayout):
(WebCore::Document::implicitClose):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayout):
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollIntoView):
(WebCore::Element::scrollIntoViewIfNeeded):
(WebCore::Element::scrollIntoViewIfNotVisible):
(WebCore::Element::scrollTo):
(WebCore::Element::scrollByUnits):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::scrollTo const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::updateCompositingLayersAfterLayout):
(WebCore::LocalFrameView::updateCompositingLayersAfterLayoutIfNeeded):
(WebCore::LocalFrameView::didLayout):
(WebCore::LocalFrameView::updateLayoutAndStyleIfNeededRecursive):
(WebCore::LocalFrameView::setTracksRepaints):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scrollingStateTreeAsText):
(WebCore::Page::setPageScaleFactor):
(WebCore::Page::layoutIfNeeded):
(WebCore::Page::updateRendering):
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/page/Page.h:
(WebCore::Page::layoutIfNeeded):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::frameViewLayoutUpdated):
(WebCore::AsyncScrollingCoordinator::frameViewEventTrackingRegionsChanged):
(WebCore::AsyncScrollingCoordinator::setEventTrackingRegionsDirty): Deleted.
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingForLayerTreeAsTextDump):
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::externalRepresentation):
(WebCore::updateLayoutIgnoringPendingStylesheetsIncludingSubframes): Deleted.

Canonical link: <a href="https://commits.webkit.org/271261@main">https://commits.webkit.org/271261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/848f91f021efc0400b5e91b880740a51a565d039

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27840 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6478 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29090 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/30112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3871 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5226 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23877 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4527 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30742 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/25433 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25342 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30881 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2863 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6229 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24634 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6682 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -339,6 +339,8 @@ enum class LayoutOptions : uint8_t {
     IgnorePendingStylesheets = 1 << 1,
     ContentVisibilityForceLayout = 1 << 2,
+    UpdateCompositingLayers = 1 << 3,
+    DontLayoutAncestorDocuments = 1 << 4,
```
Since punctuation is not available, I suggest `DoNot` instead of `Dont`, or other rewording.

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3439,6 +3448,8 @@ void Document::implicitClose()
         if (view() && renderView() && (!renderView()->firstChild() || renderView()->needsLayout()))
             protectedView()->layoutContext().layout();
+
+        protectedView()->updateCompositingLayersAfterLayoutIfNeeded();
```
The code just above checks `view()` for null, but this unconditionally dereferences `protectedView()`. Can’t both be correct.

mattwoodrow replied:
Indeed, have moved it into the same conditional (the existing update compositing layers was a subset of layout).

darinadler commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -167,6 +167,7 @@ class LocalFrameView final : public FrameView {
     void styleAndRenderTreeDidChange() override;
     bool updateCompositingLayersAfterStyleChange();
     void updateCompositingLayersAfterLayout();
+    bool updateCompositingLayersAfterLayoutIfNeeded();
```
Without a comment, I am not sure it’s obvious from the name that the `bool` return value is a “did work” indicator.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -689,11 +691,11 @@ class Document
     unsigned lastStyleUpdateSizeForTesting() const { return m_lastStyleUpdateSizeForTesting; }
 
-    WEBCORE_EXPORT void updateLayout(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
+    WEBCORE_EXPORT bool updateLayout(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
```
Without a comment, I am not sure it’s obvious from the name that the `bool` return value is a “did work” indicator.

smfr replied:
I agree that an enum return value would make the code more self-documenting.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -689,11 +691,11 @@ class Document
     // updateLayoutIgnorePendingStylesheets() forces layout even if we are waiting for pending stylesheet loads,
     // so calling this may cause a flash of unstyled content (FOUC).
-    void updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
+    bool updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
```
Without a comment, I am not sure it’s obvious from the name that the `bool` return value is a “did work” indicator.

darinadler commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1772,7 +1777,8 @@ void Page::updateRendering()
     // This is why we track m_renderingUpdateRemainingSteps as a stack.
     if (m_renderingUpdateRemainingSteps.size() > 1) {
-        layoutIfNeeded();
+        WTFLogAlways("Recursed into updateRendering, flushing and returning");
```
Do we want this? Does this fit in with our other uses of `WTFLogAlways`?

I don’t think we’d want this log to be landed unconditionally if it’s going to happen during normal operation.

mattwoodrow replied:
We definitely do not, just a remnant of my debugging process :)

smfr commented on LayoutTests/compositing/overflow/scrolling-without-painting.html:
```diff
@@ -20,17 +20,22 @@
           testRunner.dumpAsText(false);
+          testRunner.waitUntilDone();
+
+          await new Promise(resolve => requestAnimationFrame(resolve));
```
You have to be careful with this. The promise gets resolved in the microtask at the end of requestAnimationFrame callbacks, which is in the middle of `Page::updateRendering()`. Maybe you want `await UIHelper.renderingUpdate()` instead, which bounces to the next runloop?

mattwoodrow replied:
That's probably a good idea, switched to that everywhere.

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -691,11 +693,15 @@ class Document
 
-    WEBCORE_EXPORT void updateLayout(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
+    // Returns true if any style, layout or compositing layer updates
+    // were done.
```
Unwrap the comment.

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -691,11 +693,15 @@ class Document
     // so calling this may cause a flash of unstyled content (FOUC).
-    void updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> = { }, const Element* = nullptr);
+    // Returns true if any style, layout or compositing layer updates
+    // were done.
```
Unwrap the comment.

smfr commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2230,6 +2230,7 @@ void FrameLoader::commitProvisionalLoad()
             frame->protectedView()->forceLayout();
 #else
         frame->protectedView()->forceLayout();
+        frame->protectedView()->updateCompositingLayersAfterLayout();
```
Do we know why this is necessary?

mattwoodrow replied:
It would appear that it isn't, remnant of trying to track down failures.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -5027,16 +5038,18 @@ void LocalFrameView::updateLayoutAndStyleIfNeededRecursive()
     };
 
+    OptionSet<LayoutOptions> layoutOptions = { LayoutOptions::DoNotLayoutAncestorDocuments };
+    if (updateLayoutAndStyleOptions.contains(UpdateLayoutAndStyleOptions::UpdateCompositingLayers))
```
I wonder if we can share the same enum to avoid this clumsy copying.

mattwoodrow committed:
Avoid calling updateCompositingLayers() more than once before each render.
https://bugs.webkit.org/show_bug.cgi?id=84393
<rdar://11359525>

Reviewed by Simon Fraser.

Calling updateCompositingLayers after each layout causes unnecessary work, since
the layers might change again before rendering.

This change defers this work until Page::updateRendering if possible, and improves
performance on pages that force layout flushes outside of rendering.

* LayoutTests/compositing/overflow/scrolling-without-painting.html:
* LayoutTests/compositing/overflow/updating-scrolling-content.html:
* LayoutTests/compositing/scrolling/touch-scroll-to-clip.html:
* LayoutTests/compositing/updates/no-style-change-updates.html:
* LayoutTests/editing/editable-region/iframe-expected.txt:
* LayoutTests/fast/css/transform-function-perspective-crash.html:
* LayoutTests/fast/scrolling/iframe-scrollable-after-back.html:
* LayoutTests/fast/scrolling/mac/scrollbars/overflow-overlay-scrollbar-hit-test.html:
* LayoutTests/fast/scrolling/mac/scrollbars/overflow-overlay-scrollbar-hovered.html:
* LayoutTests/inspector/layers/layerTreeDidChange.html:
* LayoutTests/platform/gtk/compositing/scrolling/touch-scroll-to-clip-expected.txt: Copied from LayoutTests/platform/ios/compositing/overflow/updating-scrolling-content-expected.txt.
* LayoutTests/platform/ios/compositing/overflow/updating-scrolling-content-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/fast/text/mark-matches-overflow-clip-expected.txt:
* LayoutTests/scrollingcoordinator/scrolling-tree/clipped-layer-in-overflow-nested.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/clipped-layer-in-overflow.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-inside-relative-inside-stacking-overflow-inside-transformed.html:
* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-inside-stacking-overflow-inside-transformed.html:
* LayoutTests/tiled-drawing/scrolling/clamp-out-of-bounds-scrolls.html:
* LayoutTests/tiled-drawing/scrolling/scrolling-tree-after-scroll.html:
* LayoutTests/tiled-drawing/scrolling/slow-scrolling.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayout):
(WebCore::Document::implicitClose):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayout):
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollIntoView):
(WebCore::Element::scrollIntoViewIfNeeded):
(WebCore::Element::scrollIntoViewIfNotVisible):
(WebCore::Element::scrollTo):
(WebCore::Element::scrollByUnits):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::scrollTo const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::updateCompositingLayersAfterLayout):
(WebCore::LocalFrameView::updateCompositingLayersAfterLayoutIfNeeded):
(WebCore::LocalFrameView::didLayout):
(WebCore::LocalFrameView::updateLayoutAndStyleIfNeededRecursive):
(WebCore::LocalFrameView::setTracksRepaints):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scrollingStateTreeAsText):
(WebCore::Page::setPageScaleFactor):
(WebCore::Page::layoutIfNeeded):
(WebCore::Page::updateRendering):
(WebCore::Page::doAfterUpdateRendering):
* Source/WebCore/page/Page.h:
(WebCore::Page::layoutIfNeeded):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::frameViewLayoutUpdated):
(WebCore::AsyncScrollingCoordinator::frameViewEventTrackingRegionsChanged):
(WebCore::AsyncScrollingCoordinator::setEventTrackingRegionsDirty): Deleted.
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateCompositingForLayerTreeAsTextDump):
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::externalRepresentation):
(WebCore::updateLayoutIgnoringPendingStylesheetsIncludingSubframes): Deleted.

Canonical link: https://commits.webkit.org/271261@main
